File: backend/routes/trips.py
from fastapi import APIRouter, Depends, HTTPException ,Response ,status
from sqlalchemy.orm import Session
from datetime import date
import backend.models as models
import backend.schemamodels as schemamodels
from backend.database import get_db 
from backend import oauth2
router = APIRouter(prefix="/api", tags=["Trips"])

# for Validation
def validate_trip(vehicle, driver, cargo_weight):
    if not vehicle:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Vehicle not found.")

    if not driver:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail = "Driver not found.")

    if vehicle.status != "Available":
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail = "Vehicle is not available.")

    if driver.status != "Available":
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail = "Driver is not available.")

    if driver.license_expiry_date < date.today():
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail = "Driver license has expired.")

    if cargo_weight > vehicle.max_load_capacity_kg:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail = f"Maximum load is {vehicle.max_load_capacity_kg} kg."
        )
    

# get Users
@router.get("/trips", response_model=list[schemamodels.TripResponse])
def get_trips(
    db: Session = Depends(get_db),
    current_user: models.UserModel = Depends(oauth2.get_current_user)
):
    
    trips = db.query(models.Trip).all()

    result = []

    for trip in trips:

        result.append({

            "trip_id": trip.trip_id,

            "source": trip.source,

            "destination": trip.destination,

            "vehicle_id": trip.vehicle_id,

            "driver_id": trip.driver_id,

            "vehicle": trip.vehicle.registration_number,

            "driver": trip.driver.user.name,

            "cargo_weight_kg": trip.cargo_weight_kg,

            "planned_distance_km": trip.planned_distance_km,

            "status": trip.status,

            "created_at": trip.created_at

        })

    return result

# create New Trip
@router.post("/trips", response_model=schemamodels.TripResponse)
def create_trip(
    data: schemamodels.TripCreate,
    db: Session = Depends(get_db),
    current_user=Depends(oauth2.get_current_user)):

    # only Administrators Can Create the Trips
    if current_user.role not in ["administrators",'admin']:
        raise HTTPException(
        status_code=403,
        detail="Only administrators can Create trip."
        )
    # vehicle 
    vehicle = db.query(models.Vehicle).filter(
        models.Vehicle.vehicle_id == data.vehicle_id
    ).first()
    # driver
    driver = db.query(models.Driver).filter(
        models.Driver.driver_id == data.driver_id
    ).first()
    # Validating the Vehicle and driver
    validate_trip(vehicle=vehicle,driver=driver,cargo_weight=data.cargo_weight_kg)
    
    trip = models.Trip(

    trip_id=data.trip_id,

    source=data.source,

    destination=data.destination,

    vehicle_id=data.vehicle_id,

    driver_id=data.driver_id,

    cargo_weight_kg=data.cargo_weight_kg,

    planned_distance_km=data.planned_distance_km,

    status="Draft"
)

    db.add(trip)

    db.commit()

    db.refresh(trip)

    return {
    "trip_id": trip.trip_id,
    "source": trip.source,
    "destination": trip.destination,

    "vehicle_id": trip.vehicle_id,
    "driver_id": trip.driver_id,

    "vehicle": vehicle.registration_number,
    "driver": driver.user.name,

    "cargo_weight_kg": trip.cargo_weight_kg,
    "planned_distance_km": trip.planned_distance_km,

    "status": trip.status,

    "created_at": trip.created_at
}


# updating the Trips 
# only the Drafted Trips can be Edited , Dispatched and Completed trips cannot be edited
@router.put("/trips/{trip_id}",response_model=schemamodels.TripResponse)
def update_trip(
    trip_id: str,
    data: schemamodels.TripUpdate,
    db: Session = Depends(get_db),
    current_user: models.UserModel = Depends(oauth2.get_current_user)
):
    # only Administrators Can Create the Trips
    if current_user.role not in ["administrators",'admin']:
        raise HTTPException(
        status_code=403,
        detail="Only administrators can Edit trip."
        )
    trip = db.query(models.Trip).filter(
        models.Trip.trip_id == trip_id
    ).first()

    if not trip:
        raise HTTPException(
            status_code=404,
            detail="Trip not found."
        )
    if trip.status != "Draft":
        raise HTTPException(
            status_code=400,
            detail="Only Draft trips can be edited."
        )
    
    
    vehicle = db.query(models.Vehicle).filter(
        models.Vehicle.vehicle_id == data.vehicle_id
    ).first()
    
    
    driver = db.query(models.Driver).filter(
        models.Driver.driver_id == data.driver_id
    ).first()
    
     # Validating the Vehicle and driver
    validate_trip(vehicle=vehicle,driver=driver,cargo_weight=data.cargo_weight_kg)
    
    


    trip.source = data.source

    trip.destination = data.destination

    trip.vehicle_id = data.vehicle_id

    trip.driver_id = data.driver_id

    trip.cargo_weight_kg = data.cargo_weight_kg

    trip.planned_distance_km = data.planned_distance_km

    db.commit()

    db.refresh(trip)
    return {
        "trip_id": trip.trip_id,

        "source": trip.source,

        "destination": trip.destination,

        "vehicle_id": trip.vehicle_id,

        "driver_id": trip.driver_id,

        "vehicle": trip.vehicle.registration_number,

        "driver": trip.driver.user.name,

        "cargo_weight_kg": trip.cargo_weight_kg,

        "planned_distance_km": trip.planned_distance_km,

        "status": trip.status,

        "created_at": trip.created_at
    }




# delete Trip Api
@router.delete("/trips/{trip_id}")
def delete_trip( trip_id: str,db: Session = Depends(get_db),current_user: models.UserModel = Depends(oauth2.get_current_user)):
    # only Administrators Can Delete the Trips
    if current_user.role not in ["administrators",'admin']:
        raise HTTPException(
        status_code=403,
        detail="Only administrators can Delete trip."
        )
    trip = db.query(models.Trip).filter(
        models.Trip.trip_id == trip_id
    ).first()

    if not trip:
        raise HTTPException(
            status_code=404,
            detail="Trip not found."
        )

    if trip.status != "Draft":
        raise HTTPException(
            status_code=400,
            detail="Only Draft trips can be deleted."
        )
    
    db.delete(trip)
    db.commit()
    return {''}




# dispatch Api
@router.patch("/trips/{trip_id}/dispatch")
def dispatch_trip(
    trip_id: str,
    db: Session = Depends(get_db),
    current_user: models.UserModel = Depends(oauth2.get_current_user)
):
     # only Administrators Can Delete the Trips
    if current_user.role not in ["administrators",'admin']:
        raise HTTPException(
        status_code=403,
        detail="Only administrators can Dispatch trip."
        )
    trip = db.query(models.Trip).filter(
        models.Trip.trip_id == trip_id
    ).first()

    if not trip:
        raise HTTPException(404, "Trip not found.")

    if trip.status != "Draft":
        raise HTTPException(400, "Only Draft trips can be dispatched.")

    vehicle = db.query(models.Vehicle).filter(
        models.Vehicle.vehicle_id == trip.vehicle_id
    ).first()

    
    driver = db.query(models.Driver).filter(
        models.Driver.driver_id == trip.driver_id
    ).first()

    validate_trip(vehicle, driver, trip.cargo_weight_kg)

    trip.status = "Dispatched"
    vehicle.status = "On Trip"
    driver.status = "On Trip"

    db.commit()
    db.refresh(trip)

    return {
        "message": "Trip dispatched successfully.",
        "trip_id": trip.trip_id,
        "status": trip.status
    }



# complete Api
@router.patch("/trips/{trip_id}/complete")
def complete_trip(
    trip_id: str,
    db: Session = Depends(get_db),
    current_user: models.UserModel = Depends(oauth2.get_current_user)
):
     # only Administrators Can Delete the Trips
    if current_user.role not in ["administrators",'admin']:
        raise HTTPException(
        status_code=403,
        detail="Only administrators can Dispatch trip."
        )
    trip = db.query(models.Trip).filter(
        models.Trip.trip_id == trip_id
    ).first()

    if not trip:
        raise HTTPException(
            status_code=404,
            detail="Trip not found."
        )

    if trip.status != "Dispatched":
        raise HTTPException(
            status_code=400,
            detail="Only dispatched trips can be completed."
        )

    vehicle = db.query(models.Vehicle).filter(
        models.Vehicle.vehicle_id == trip.vehicle_id
    ).first()

    if not vehicle:
        raise HTTPException(
            status_code=404,
            detail="Vehicle not found."
        )

    driver = db.query(models.Driver).filter(
        models.Driver.driver_id == trip.driver_id
    ).first()

    if not driver:
        raise HTTPException(
            status_code=404,
            detail="Driver not found."
        )

    trip.status = "Completed"

    vehicle.status = "Available"

    driver.status = "Available"

    db.commit()

    db.refresh(trip)

    return {
        "message": "Trip completed successfully.",
        "trip_id": trip.trip_id,
        "status": trip.status
    }

File: frontend/staticFiles/trips.js
let tripsCache = [];
let driversMap = {}; // Maps driver_id -> driver details
let vehiclesMap = {}; // Maps vehicle_id -> vehicle details

async function loadDropdowns() {
  try {
    //-------------------------
    // Drivers
    //-------------------------

    const driverResponse = await fetch(`${host}/api/drivers/available`, {
      credentials: "include",
    });

    if (!driverResponse.ok) {
      throw new Error("Unable to load drivers.");
    }

    const drivers = await driverResponse.json();

    const driverSelect = document.getElementById("f-driver");

    driverSelect.innerHTML = `<option value="">Select Driver</option>`;

    drivers.forEach((driver) => {
      driversMap[driver.driver_id] = driver;
      driverSelect.innerHTML += `
                <option value="${driver.driver_id}">
                ${driver.name} (${driver.contact_number})
            </option>
            `;
    });

    //-------------------------
    // Vehicles
    //-------------------------

    const vehicleResponse = await fetch(`${host}/api/vehicles/available`, {
      credentials: "include",
    });

    if (!vehicleResponse.ok) {
      throw new Error("Unable to load vehicles.");
    }

    const vehicles = await vehicleResponse.json();

    const vehicleSelect = document.getElementById("f-vehicle");

    vehicleSelect.innerHTML = `<option value="">Select Vehicle</option>`;

    vehicles.forEach((vehicle) => {
      vehiclesMap[vehicle.vehicle_id] = vehicle;

      vehicleSelect.innerHTML += `
                <option value="${vehicle.vehicle_id}">
                    ${vehicle.registration_number}
                    (${vehicle.name_model})
                    (${vehicle.max_load_capacity_kg} kg)
                </option>
            `;
    });
  } catch (err) {
    console.error(err);

    showToast(err.message, "error");
  }
}
function escapeHTML(str) {
  if (!str) return "";
  return String(str).replace(
    /[&<>'"]/g,
    (t) =>
      ({
        "&": "&amp;",
        "<": "&lt;",
        ">": "&gt;",
        "'": "&#39;",
        '"': "&quot;",
      })[t] || t,
  );
}

// load Data
async function loadData() {
  try {
    const response = await fetch(`${host}/api/trips`, {
      credentials: "include",
    });

    if (!response.ok) {
      throw new Error("Unable to load trips.");
    }

    tripsCache = await response.json();

    renderTable(tripsCache);

    renderStats(tripsCache);
  } catch (err) {
    console.error(err);

    showToast(err.message, "error");
  }
}

function renderStats(data) {
  document.getElementById("stat-total").textContent = data.length;
  document.getElementById("stat-dispatched").textContent = data.filter(
    (i) => i.status === "Dispatched",
  ).length;
  document.getElementById("stat-completed").textContent = data.filter(
    (i) => i.status === "Completed",
  ).length;

  const activeCargo = data
    .filter((i) => i.status === "Dispatched")
    .reduce((acc, curr) => acc + curr.cargo_weight_kg, 0);
  document.getElementById("stat-cargo").textContent =
    `${(activeCargo / 1000).toFixed(1)} T`;
}

function renderTable(data) {
  const tbody = document.getElementById("table-body");
  tbody.innerHTML = "";

  if (data.length === 0) {
    tbody.innerHTML = `
            <tr>
                <td colspan="6" class="p-8 text-center text-slate-400">
                    No active trips found.
                </td>
            </tr>
        `;
    return;
  }

  data.forEach((trip) => {
    let badgeClass = "bg-slate-100 text-slate-600 border-slate-300";
    let badgeDot = "bg-slate-400";

    switch (trip.status) {
      case "Dispatched":
        badgeClass = "bg-sky-50 text-sky-700 border-sky-200";
        badgeDot = "bg-sky-500 animate-pulse";
        break;

      case "Completed":
        badgeClass = "bg-emerald-50 text-emerald-700 border-emerald-200";
        badgeDot = "bg-emerald-500";
        break;

      case "Cancelled":
        badgeClass = "bg-rose-50 text-rose-700 border-rose-200";
        badgeDot = "bg-rose-500";
        break;
    }

    const driver = driversMap[trip.driver_id];
    const vehicle = vehiclesMap[trip.vehicle_id];

    const driverName = driver ? driver.name : `Driver #${trip.driver_id}`;

    const driverLicense = driver ? driver.license_number : "N/A";

    const vehicleReg = vehicle
      ? vehicle.registration_number
      : `Vehicle #${trip.vehicle_id}`;

    const vehicleModel = vehicle ? vehicle.name_model : "";

    const tr = document.createElement("tr");
    tr.className = "hover:bg-slate-50 transition-colors group";
    let actions = "";

    if (trip.status === "Draft") {
      actions = `

        <div class="flex items-center justify-center gap-2">

            <button
                class="edit-btn p-2 rounded-lg hover:bg-sky-50 text-slate-500 hover:text-sky-600"
                data-id="${trip.trip_id}">

                <i data-lucide="pencil" class="w-4 h-4"></i>

            </button>

            <button
                class="delete-btn p-2 rounded-lg hover:bg-rose-50 text-slate-500 hover:text-rose-600"
                data-id="${trip.trip_id}">

                <i data-lucide="trash-2" class="w-4 h-4"></i>

            </button>

            <button
                class="dispatch-btn px-3 py-1 rounded-full
                bg-emerald-500 text-white text-xs font-semibold
                hover:bg-emerald-600"
                data-id="${trip.trip_id}">

                Dispatch

            </button>

        </div>

    `;
    } else if (trip.status === "Dispatched") {
      actions = `

        <button
            class="complete-btn
            px-3 py-1
            rounded-full
            bg-sky-600
            text-white
            text-xs
            font-semibold"
            data-id="${trip.trip_id}">

            Complete

        </button>

    `;
    } else {
      actions = `

        <span class="text-slate-400 text-xs">

            Completed

        </span>

    `;
    }
    tr.innerHTML = `

            <td class="px-6 py-4 font-mono font-semibold text-xs text-slate-900">
                ${escapeHTML(trip.trip_id)}
            </td>

            <td class="px-6 py-4">
                <div class="flex flex-col gap-1">

                    <div class="flex items-center gap-2 text-slate-800 text-sm font-medium">
                        <i data-lucide="map-pin" class="w-4 h-4 text-green-600"></i>
                        ${escapeHTML(trip.source)}
                    </div>

                    <div class="flex items-center gap-2 text-slate-500 text-sm">
                        <i data-lucide="move-right" class="w-4 h-4"></i>
                        ${escapeHTML(trip.destination)}
                    </div>

                </div>
            </td>

            <td class="px-6 py-4">
                <div class="space-y-3">

                    <div class="flex items-start gap-2">
                        <i data-lucide="truck" class="w-4 h-4 mt-0.5 text-sky-600"></i>

                        <div>
                            <div class="text-sm font-semibold text-slate-800">
                                ${escapeHTML(vehicleReg)}
                            </div>

                            <div class="text-xs text-slate-500">
                                ${escapeHTML(vehicleModel)}
                            </div>
                        </div>
                    </div>

                    <div class="flex items-start gap-2">
                        <i data-lucide="user" class="w-4 h-4 mt-0.5 text-emerald-600"></i>

                        <div>
                            <div class="text-sm font-semibold text-slate-800">
                                ${escapeHTML(driverName)}
                            </div>

                            <div class="text-xs text-slate-500">
                                ${escapeHTML(driverLicense)}
                            </div>
                        </div>
                    </div>

                </div>
            </td>

            <td class="px-6 py-4">

                <div class="text-sm font-semibold text-slate-800">
                    ${Number(trip.cargo_weight_kg).toLocaleString()} kg
                </div>

                <div class="text-xs text-slate-500 mt-1">
                    ${Number(trip.planned_distance_km).toLocaleString()} km
                </div>

            </td>

            <td class="px-6 py-4">

                <span class="inline-flex items-center gap-2 px-3 py-1 rounded-full border text-xs font-semibold ${badgeClass}">

                    <span class="w-2 h-2 rounded-full ${badgeDot}"></span>

                    ${escapeHTML(trip.status)}

                </span>

            </td>
            <td class="px-6 py-4 text-center">

               ${actions}

            </td>

            
        `;

    tbody.appendChild(tr);
  });

  tbody.querySelectorAll(".edit-btn").forEach((btn) => {
    btn.addEventListener("click", () => openModal(btn.dataset.id));
  });

  tbody.querySelectorAll(".delete-btn").forEach((btn) => {
    btn.addEventListener("click", () => confirmDelete(btn.dataset.id));
  });
  tbody.querySelectorAll(".dispatch-btn").forEach((btn) => {
    btn.addEventListener("click", () => {
      confirmDispatch(btn.dataset.id);
    });
  });
  tbody.querySelectorAll(".complete-btn").forEach((btn) => {
    btn.addEventListener("click", () => {
      confirmComplete(btn.dataset.id);
    });
  });
  lucide.createIcons();
}

const modal = document.getElementById("modal");
const form = document.getElementById("data-form");

function openModal(id = null) {
  form.reset();
  modal.classList.remove("hidden");
  const idInput = document.getElementById("f-trip-id");

  if (id) {
    const t = tripsCache.find((i) => i.trip_id === id);
    document.getElementById("modal-title").textContent = "Edit Trip Dispatch";
    document.getElementById("form-is-edit").value = "true";

    idInput.value = t.trip_id;
    idInput.readOnly = true; // Cannot edit PK

    document.getElementById("f-source").value = t.source;
    document.getElementById("f-dest").value = t.destination;
    document.getElementById("f-vehicle").value = t.vehicle_id;
    document.getElementById("f-driver").value = t.driver_id;
    document.getElementById("f-cargo").value = t.cargo_weight_kg;
    document.getElementById("f-dist").value = t.planned_distance_km;
    document.getElementById("f-status").value = t.status;
  } else {
    document.getElementById("modal-title").textContent = "Dispatch New Trip";
    document.getElementById("form-is-edit").value = "false";

    // Generate aesthetic random Trip ID TRP-XXXX
    idInput.value =
      "TRP-" +
      Date.now().toString().slice(-6) +
      "-" +
      Math.floor(Math.random() * 1000);
    idInput.readOnly = false;
  }
}

function closeModal() {
  modal.classList.add("hidden");
}
document
  .getElementById("add-trip-btn")
  .addEventListener("click", () => openModal());
document
  .getElementById("close-modal-btn")
  .addEventListener("click", closeModal);
document
  .getElementById("cancel-modal-btn")
  .addEventListener("click", closeModal);

function confirmDelete(id) {
  showConfirmModal({
    title: "Delete Trip",

    message: "Are You sure To Delete This trip? This can`t Be back.",

    confirmText: "Delete",

    confirmClass: "bg-red-600 hover:bg-red-700",

    onConfirm: async () => {
      await deleteData(id);
    },
  });
}
async function deleteData(id) {
  try {
    const res = await fetch(`${host}/api/trips/${id}`, {
      method: "DELETE",
      credentials: "include",
    });
    if (!res.ok) throw new Error("Delete failed");

    showToast("Trip cancelled/deleted", "success");
    await loadData();
    await loadDropdowns();
  } catch (err) {
    console.error(err);
    showToast(err.message, "error");
  }
}

form.addEventListener("submit", async (e) => {
  e.preventDefault();
  const isEdit = document.getElementById("form-is-edit").value === "true";
  const trip_id = document.getElementById("f-trip-id").value.trim();

  const payload = {
    trip_id,
    source: document.getElementById("f-source").value.trim(),
    destination: document.getElementById("f-dest").value.trim(),
    vehicle_id: parseInt(document.getElementById("f-vehicle").value),
    driver_id: parseInt(document.getElementById("f-driver").value),
    cargo_weight_kg: parseFloat(document.getElementById("f-cargo").value),
    planned_distance_km: parseFloat(document.getElementById("f-dist").value),
  };

  try {
    const url = isEdit ? `${host}/api/trips/${trip_id}` : `${host}/api/trips`;
    const res = await fetch(url, {
      method: isEdit ? "PUT" : "POST",
      credentials: "include",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });
    if (!res.ok) {
      // 1. Try to parse the server's JSON error message
      const errorData = await res.json().catch(() => ({}));

      // 2. Use the server message, fallback to status text, or a default string
      const errorMessage =
        errorData.detail || res.statusText || "API Sync Failed";

      throw new Error(errorMessage);
      // const errorData = await res.json().catch(() => ({}));

      // let errorMessage = "API Sync Failed";

      // if (Array.isArray(errorData.detail)) {
      //   errorMessage = errorData.detail
      //     .map((err) => `${err.loc.join(" -> ")} : ${err.msg}`)
      //     .join("\n");
      // } else if (typeof errorData.detail === "string") {
      //   errorMessage = errorData.detail;
      // }

      // throw new Error(errorMessage);
    }
    await loadData();

    closeModal();

    showToast(
      isEdit ? "Trip updated successfully" : "Trip created successfully",
      "success",
    );
  } catch (err) {
    console.error(err);
    showToast(err.message, "error");
  }
});

function confirmDispatch(id) {
  showConfirmModal({
    title: "Dispatch Trip",

    message: `Dispatch Trip ${id}?`,

    confirmText: "Dispatch",

    confirmClass: "bg-emerald-600 hover:bg-emerald-700",

    onConfirm: async () => {
      await dispatchTrip(id);
    },
  });
}
async function dispatchTrip(id) {
  try {
    const response = await fetch(
      `${host}/api/trips/${id}/dispatch`,

      {
        method: "PATCH",

        credentials: "include",
      },
    );

    const data = await response.json();

    if (!response.ok) {
      throw new Error(data.detail || "Dispatch Failed");
    }

    showToast("Trip Dispatched Successfully", "success");

    await loadDropdowns();

    await loadData();
  } catch (err) {
    console.error(err);

    showToast(err.message, "error");
  }
}
function confirmComplete(id) {
  showConfirmModal({
    title: "Complete Trip",

    message: `Complete Trip ${id}?`,

    confirmText: "Complete",

    confirmClass: "bg-sky-600 hover:bg-sky-700",

    onConfirm: async () => {
      await completeTrip(id);
    },
  });
}
async function completeTrip(id) {
  try {
    const response = await fetch(
      `${host}/api/trips/${id}/complete`,

      {
        method: "PATCH",
        credentials: "include",
      },
    );

    const data = await response.json();

    if (!response.ok) {
      throw new Error(data.detail || "Completion Failed");
    }

    showToast("Trip Completed Successfully", "success");

    await loadDropdowns();

    await loadData();
  } catch (err) {
    console.error(err);

    showToast(err.message, "error");
  }
}
